### ISSUE TITLE
[EKS] [request]: Control Plane Logs to S3

### ISSUE THREAD

**Issue by mike-stewart** [aws/containers-roadmap#1141]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Tell us about your request**
EKS currently supports control plane logging to CloudWatch Logs (#26), but it would be helpful to support S3 as an alternative log destination. This would be in line with many other AWS services that support logging both to both CloudWatch and S3.

**Which service(s) is this request for?**
EKS

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
Enabling S3 as a log destination would help with a few issues.

**_Cost_**: CloudWatch Logs is much more expensive than S3 for both ingestion and storage of logs, and it can easily cost in the hundreds of dollars a month to ingest the control plane logs even for a small cluster. In my experience, the **_cost of CloudWatch Log ingestion for many clusters exceeds the cost of EKS itself_**. For customers that have compliance requirements that require the collection of audit logs but do not need all the features of Cloudwatch Logs, archiving logs to S3 would be much more economical.

**_Flexibility_**: Archiving logs to S3 would allow customers that don't use Cloudwatch Logs more flexibility, such as Athena or cost-effective ingestion into another tool like AWS Elasticsearch or Splunk.

**Are you currently working around this issue?**
Ingesting logs through CloudWatch logs, at a much higher cost.

**Comment by pparth** [comment]
@tabern  is there an update here? As @mike-stewart  mentioned, the cost of CloudWatch log ingestion for big clusters is prohibiting, especially when we need to have access to audit logs. We need to have the logs directly published to an S3 bucket for further processing.

**Comment by StealthyDev** [comment]
is there a workaround to this?

**Comment by aravindh-seeneevasan** [comment]
Please prioritise this. This definitely saves cost for lots of customers across the globe

**Comment by youssemn** [comment]
is there any update about this issue ?

**Comment by eightnoteight** [comment]
@tabern please prioritise this. currently our eks cluster is ingesting about 40TB per month which is a huge waste of cost for us. such cost issues are prohibiting us from moving our other big workloads from ecs/ec2 to eks.

**Comment by debu99** [comment]
Any update?

**Comment by dogzzdogzz** [comment]
I created a feature request support ticket and my TAM told me that this feature is on planning roadmap to be released in 12+ months, hopefully this feature can be prioritized because EKS logs grows quite fast with our cluster scale keeping growing and the cloudwatch is much more expensive than s3 for storing logs

**Comment by kderck** [comment]
This would be great.

**Comment by jesseadev** [comment]
@tabern Can we push this sooner?

**Comment by jsokol89** [comment]
not gonna lie this would be great to have!

**Comment by aravindh-seeneevasan** [comment]
We currently storing the logs to s3 via cloudwatchlog -> lambdasubscription(process logs) ->s3

Note: Log group retention set to 5 days for reducing the cost

**Comment by dogzzdogzz** [comment]
@aravindh-seeneevasan The data ingestion cost of cloudwatch log is much more expensive than the storage, moving log to from cloudwatch to s3 can not save much.

**Comment by zeppelinen** [comment]
Any known workarounds? We need audit logs but their costs are enormous.

**Comment by ClaytonOlleyNutrien** [comment]
Cloudwatch is far too expensive for these kinds of logs that need to be fed into our SIEM anyway. Please give us a direct to s3 option!

**Comment by JeffreyRaleighHall-KR** [comment]
As Einstein said: "Everything should be made as simple as possible, but no simpler".   EKS may be unusable for us until it has ability to deliver Audit logs without needing Cloudwatch.

**Comment by RichardoC** [comment]
> Any known workarounds? We need audit logs but their costs are enormous.

If you only need mutations you can use <https://github.com/RichardoC/kube-audit-rest>

Unfortunately if you need to log read only actions, there's no other option at this time.

**Comment by RichardoC** [comment]
> Any known workarounds? We need audit logs but their costs are enormous.

If you (@zeppelinen) only need mutations you can use <https://github.com/RichardoC/kube-audit-rest>

Unfortunately if you need to log read only actions, there's no other option at this time.

**Comment by etendards** [comment]
I heard this feature is not even on the EKS roadmap.  We have made several requests for this feature.

**Comment by max-tremblay** [comment]
As a cost mitigation solution, AWS released the  infrequent access logs option in CloudWatch last November, which aims to reduce costs by half for those who do not require any advanced features:

https://aws.amazon.com/fr/blogs/aws/new-amazon-cloudwatch-log-class-for-infrequent-access-logs-at-a-reduced-price/

**Comment by moebaca** [comment]
Following up on @max-tremblay 's comment - delivery to CloudWatch Logs Infrequent Access is probably as good as it's going to get for ingest cost optimization. Even if AWS adds the support to export Control Plane vended logs directly to S3 it will likely cost exactly the same as CW Infrequent Access as the [pricing page](https://aws.amazon.com/cloudwatch/pricing/) illustrates:
<img width="1160" alt="image" src="https://github.com/aws/containers-roadmap/assets/12791848/6189870b-1ab8-4d47-ab68-fad4e57d95c4">

**Comment by joebowbeer** [comment]
@moebaca According to the table you reference, Infrequent Access does not support storing in S3.

IOW, those needing audit logs in S3, e.g., for 3rd party integrations, will not be able to benefit from the savings provided by Infrequent Access.

**Comment by moebaca** [comment]
@joebowbeer definitely. I was specifically calling out cost optimization since there are many of us here strictly concerned with that. This PR is definitely still relevant outside, but for those of us here just for saving on log ingest fees alone.. well we pretty much bottomed out now with the new storage class since ingest into S3 will likely cost the same.

**Comment by jfarrell** [comment]
@moebaca this is incorrect, the table you reference is for Cloudwatch to push logs outbound after ingestion and that cost shown in that table is for you to pay the added push, not a comparison. 

S3 ingest is [$0.00](https://aws.amazon.com/s3/pricing)

Additionally EKS does not expose the ability to set the CloudWatch Storage Class when creating the log group and CloudWatch does not allow for the storage class to be modified after creation. 

EKS Team we need some help here with either S3 as a push option or exposing a way to tune the CloudWatch Storage Class or provide which CloudWatch log group to use as a configuration so it can be created with whatever settings we want.

**Comment by techmouse84** [comment]
It's sad to see that this is still an issue after so many years. My guess is that it affects AWS bottom line. 
Anyway, Aws Security Lake offers to get ur audit logs for you to s3 at the same price as cloudwatch IA. 
Just need to be aware there's lambda cost as well.

**Comment by MartinEmrich** [comment]
I just contacted our TAM team about this topic, I suggest everyone here does the same.

**Comment by BhuviTheDataGuy** [comment]
+ 1

**Comment by vaibhavhirani** [comment]
Atleast they could have given us an option to switch the log class to Infrequent. 

Was looking into third party solution such as Fluentbit and functionbeat as well but these are not coming through. AWS already uses Fluentbit and functionbeat relies on cloudwatch so it get knocked out of the possible very easily.

Anyone tried fluentbit to get logs from control plane in eks? although I doubt that control place will be accessible to such components.

**Comment by RichardoC** [comment]
> Atleast they could have given us an option to switch the log class to Infrequent.
> 
> Was looking into third party solution such as Fluentbit and functionbeat as well but these are not coming through. AWS already uses Fluentbit and functionbeat relies on cloudwatch so it get knocked out of the possible very easily.
> 
> Anyone tried fluentbit to get logs from control plane in eks? although I doubt that control place will be accessible to such components.

The control plane logs are not directly accessible, you either have to get them via cloudwatch or a bodge like [kube-audit-rest](https://github.com/RichardoC/kube-audit-rest)

Any other tools are just consuming those cloudwatch logs and sending them somewhere else. Here's an example from Falco <https://falco.org/blog/k8saudit-eks-plugin/>

**Comment by tropnikovvl** [comment]
@RichardoC 
Hello!

Do I understand correctly that [kube-audit-rest](https://github.com/RichardoC/kube-audit-rest) collects API server logs, but does not collect audit and authentication logs?

**Comment by RichardoC** [comment]
> @RichardoC Hello!
> 
> Do I understand correctly that [kube-audit-rest](https://github.com/RichardoC/kube-audit-rest) collects API server logs, but does not collect audit and authentication logs?

It captures all mutation/creation API calls, so any requests that pass authentication and authorisation subject to the [limitations](https://github.com/RichardoC/kube-audit-rest?tab=readme-ov-file#known-limitations-and-warnings)

**Comment by lifttocode** [comment]
> We currently storing the logs to s3 via cloudwatchlog -> lambdasubscription(process logs) ->s3 -> fanout to multiple destinations via SNS
> 
> Note: Log group retention set to 5 days for reducing the cost

I believe this approach is highly valid, and it aligns with the recommendations in the EKS best practice guide [Stream Logs to S3](https://aws.github.io/aws-eks-best-practices/cost_optimization/cost_opt_observability/#stream-logs-to-s3), which suggests a similar setup. I think it’s worth trying out.

That said, if anyone disagrees with this method, could you share the reasoning? Is it because of potential additional costs incurred by the pipeline setup, or are there other concerns?

**Comment by dogzzdogzz** [comment]
@lifttocode [This](https://github.com/aws/containers-roadmap/issues/1141#issuecomment-1271028724) is the reason

**Comment by hver** [comment]
There is a new option to save the EKS logs in S3: SecurityLake. Still you have to pay extra for SecurityLake to ingest and normalize logs, but in the end it is cheaper than CloudWatch -> S3.

**Comment by MartinEmrich** [comment]
@hver At least here on my test cluster (eu-central-1, EKS 1.31), there's no visible option like a "Switch to security lake".  So if you enable Control Plane logs, the horrendous cost of CloudWatch logs is incurred.

SecurityLake seems to be an addition to this (most probably it just taps into CloudWatch Logs behind the scenes). It's cost will not replace CloudWatch, but add to it.

The main goal of this issue is (at least for me, and I assume for many of us), to prevent the high cost of CloudWatch, when we already run another log aggregation service (ELK, Grafana Loki, Graylog, Splunk, DataDog, whatever). So all we need is a way to get the logs to pipe there with minimal, reaonable cost.

**Comment by lifttocode** [comment]
> @lifttocode [This](https://github.com/aws/containers-roadmap/issues/1141#issuecomment-1271028724) is the reason

alright, i totally missed that – it’s not cheap at all

**Comment by hver** [comment]
@MartinEmrich SecurityLake is designed for centralized security configuration, so you have to turn it on SecurityLake page, and activate gathering of eks logs. As I previously wrote, SecurityLake is not using CloudWatch but independent AWS channels(https://docs.aws.amazon.com/eks/latest/userguide/integration-securitylake.html) . You should disable EKS Audit logs in EKS (CloudWatch) to avoid duplicated costs.
Then you can use Athena to search in SecurityLake logs. 
Still I would prefer to have an option for saving EKS in S3, but at least with SecurityLake we were able to cut the costs significantly

**Comment by MartinEmrich** [comment]
@hver I am still not sure about that. The docs explicitly say "enable control plane logs" https://docs.aws.amazon.com/eks/latest/userguide/integration-securitylake.html. While it won't consume the logs through customer-facing channels (a customer or attacker could sabotage), I understand that sending the logs from EKS is still required... Though I have not tested it, maybe we'll give Security Lake a spin next year.

But this seems specifically about *Audit* Logs. That's similar to AWS GuardDuty, which also consumes EKS Audit Logs, even if "Control, Plane Logs" is not enabled. (And we tried it recently, that definitively works without Control Plane Logs).

I would be more interested in the "non"-audit-Logs (Kube-Controller-Manager, etc.), to analyze stuff like pod scheduling issues, etc.

**Comment by henkka** [comment]
We discussed this with AWS folks and received information from them that the EKS Audit Logs to Security Lake integration does not require any CWLogs integration, and thus, the ingestion costs for the EKS Audit Logs is solely the AWS Security Lake ingestion and normalization costs.

**Comment by MartinEmrich** [comment]
> the ingestion costs for the EKS Audit Logs is solely the AWS Security Lake ingestion and normalization costs.

Makes sense, as AWS GuardDuty (processing EKS _audit_ logs directly) works the same way.

(Does not solve the regular control plane logs to S3 though)

**Comment by maradwan** [comment]
It's surprising that we're in 2025 and this feature still hasn't been implemented. I'm running more than 40 EKS clusters, and the logging costs have become extremely expensive.

**Comment by claytonolley** [comment]
I see other services offering other logging destinations recently so perhaps this one will get that eventually too. Even with the vended logs discount, CloudWatch seems far more expensive than its worth.

**Comment by weitzj** [comment]
Only Solution Right now seems to use AWS security lake I guess .

Then if you don’t store your logs in AWS , get a partner which joined the AWS security lake partner program.

Idea: aws has to have the logs already before  you say you want to store it in cloudwatch or S3

**Comment by aspx** [comment]
> It's surprising that we're in 2025 and this feature still hasn't been implemented. I'm running more than 40 EKS clusters, and the logging costs have become extremely expensive.

I completely agree - it’s honestly shocking that in 2025 we still can’t natively stream eks audit logs directly to S3..! We're seeing similar issues across our clusters, in some cases, the cost of audit logging exceeds the actual compute resources. Pending in the backlog for 5y🫣

**Comment by maradwan** [comment]
If you use the GuardDuty service, you'll notice an increase in the "PaidKubernetesAuditLogsAnalyzed" charges.

**Comment by bowale01** [comment]
I hope in 2026 they will work on this.

**Comment by zweizhang** [comment]
I implemented the approach described below to reduce CloudWatch PutLogEvents costs and achieved a significant cost reduction. This solution meets audit requirements and cuts costs by half, as the Infrequent Access (IA) ingest pricing is 50% lower than standard ingest.

For reference:
https://repost.aws/zh-Hans/articles/ARsjlEAAI0RGinbtSeTDx1kA/optimizing-costs-with-the-cloudwatch-logs-infrequent-access-log-class-for-amazon-eks-control-plane-logging

**Comment by nantiferov** [comment]
**Infrequent Access (IA)** logs works fine, but afaik with them there's no way to deliver these logs to external storage, like Elastic or Loki due to API limitations.

**Comment by max-tremblay** [comment]
@nantiferov it wasn't the case when they released that feature, but now S3 export is possible:

https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/CloudWatch_Logs_Log_Classes.html

**Comment by maradwan** [comment]
The main issue is the cost of VendedLog-Bytes in CloudWatch Logs, which comes from EKS control plane logs. Currently, EKS can only send these logs to CloudWatch, and there is still no option to send them directly to S3.

This makes audit logging quite expensive, especially for clusters with high API activity.

**Comment by MartinEmrich** [comment]
> Currently, EKS can only send these logs to CloudWatch, and there is still no option to send them directly to S3.

Yes, and that is the core point of this issue. And, hot take... The more "workarounds" and "hacks" actually involving costly CloudWatch logs are posted here, the more this might signal AWS that the incentive for actually implementing direct-to-S3 logging is not that high. (Will nag our TAM again today about this issue).

**Comment by igor-nikiforov** [comment]
VendedLog-Bytes in CW cost us almost 700$ just for audit events. It's ridiculous that this is still an issue after almost 6 years.
